src/main/java/de/tum/in/www1/artemis/service/MailService.java

```diff
@@ -153,6 +153,48 @@ public void sendSAML2SetPasswordMail(User user) {
 
     // notification related
 
+    /**
+     * Sets the correct context for the case that the notificationSubject is an Exercise
+     * @param notificationSubject which is an Exercise
+     * @param context that is modified based on the exercise type
+     */
+    private void setExerciseContext(Object notificationSubject, Context context) {
+        if (notificationSubject instanceof QuizExercise) {
+            context.setVariable(EXERCISE_TYPE, "quiz");
+        }
+        else if (notificationSubject instanceof ModelingExercise) {
+            context.setVariable(EXERCISE_TYPE, "modeling");
+        }
+        else if (notificationSubject instanceof TextExercise) {
+            context.setVariable(EXERCISE_TYPE, "text");
+        }
+        else if (notificationSubject instanceof ProgrammingExercise) {
+            context.setVariable(EXERCISE_TYPE, "programming");
+        }
+        else if (notificationSubject instanceof FileUploadExercise) {
+            context.setVariable(EXERCISE_TYPE, "upload");
+        }
```

Comment:
[nit] What do you think of introducing `getExerciseType()` method on the Exercise class and then provide the trivial implementation for it, as in:
```java
enum ExerciseType {
  case TEXT, MODELING, ...
}

class TextExercise {
  ExerciseType getExerciseType() {
    return ExerciseType.TEXT;
  }
}
```

We can make this method (`getExerciseType`) as required one and then if new Exercise is introduced, this code logic (`setExerciseContext`) will just work without any extra though.

I guess otherwise it makes sense to at least log (or assert) in the `else` case, as we won't set the exercise for the exercise which will hit the else case.

PS. Oh, just noticed it's not your code.

Reply:
yes, same applies to the other comments. I think it's better to address this in a mail/notification related PR. But thanks for the input :)

Reply:
Very good idea. I simply crafted something more or less on the fly. I added your change to a follow-up PR (see my answer to your other comment about the MailService for more info)